### ISSUE TITLE
Need API similar to model.setProperties that would return a promise

### ISSUE THREAD

**Issue by lolmaus** [orbitjs/ember-orbit#171]
# TL/DR

Introduce a new `record.update()` method that lets you set keys, attributes and relationships in one call, similar to `record.setProperties()`, but also returns a promise.

# Use case

We've been using `record.setProperties({attr1, attr2, rel1, rel2})` in a controller. It resulted in tests failing *sometimes*, unpredictably. When by lucky chance we had the failure consistently reproducible on a laptop, we were able to debug it.

The cause of the problem is  that `setProperties` implicitly runs a background server interaction. It was not waited for, the test was completed and torn down. The response returned before the next test was initialized, crashing our WebSocket mock.

In order to properly work around the issue, we had to change this:

```js
  doggy = null; // populated with a 

  @action
  edit(attr1, attr2, rel1, rel2) {
    this.doggy.setProperties({attr1, attr2, rel1, rel2});
  }
```

to this:

```js
  @waitFor // a decorator that makes tests aware of a pending operation
  updateDoggy(attr1, attr2, rel1, rel2) {
    return this.store.update((t) => [
      t.replaceAttribute(this.doggy.identity, 'attr1', attr1),
      t.replaceAttribute(this.doggy.identity, 'attr2', attr2),
     t.replaceRelatedRecord(this.doggy.identity, 'rel1', rel1 && rel1.identity),
      t.replaceRelatedRecord(this.doggy.identity, 'rel2', rel2 && rel2.identity)
    ]);
  }

  @action
  edit(attr1, attr2, rel1, rel2) {
    this.updateDoggy(attr1, attr2, rel1, rel2);
  }
```

It works, but it is:

* too much boilerplate code for a simple `setProperties`
* too tightly coupled to a specific use case

Instead, there should be a generic `setProperties` equivalent that returns a promise.

# Suggested implementation

1. Have all `replace*` methods in [model.js](https://github.com/orbitjs/ember-orbit/blob/v0.15.9/addon/-private/model.js) return promises. It's necessary for consistency: currently only `replaceAttributes` returns a promise.
2. Create a plural version of `replaceRelatedRecord` to be consistent with `replaceAttribute`/`replaceAttributes`.
3. Create a new `update` method on the model that has API similar to `setProperties` but returns a promise.

This new `update` method must be able to distinguish keys, attributes and relationships. This is necessary because keys, attributes and relationships can only be updated with separate methods (unless you use `setProperties` which doesn't return a promise and triggers multiple `store.update()`s). This can be achieved by looking into `Model.keys`, `Model.attributes` and `Model.relationships`. Arguments that are not keys/attributes/relationships can be ignored.

# Notes

I'm eager to start a PR. Tell me if you agree with the described path.

**Comment by dgeb** [comment]
@lolmaus I must confess that I haven't run into these issues because I normally make edits in store forks and then merge them back. However, this looks like a great alternative approach. Please proceed with your PR and let me know if you have any questions as you go. Thanks in advance.

**Comment by tchak** [comment]
@lolmaus I added to my store (you can do it on the model) a method like this :

```js
  async sync() {
    await this.source.requestQueue.process();
    await this.source.syncQueue.process();
  }
```

**Comment by dgeb** [comment]
Implemented via #188
